plugins/synaptics-rmi/fu-synaptics-rmi-device.c

```diff
@@ -0,0 +1,1098 @@
+				     guint datasz,
+				     GError **error)
+{
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+
+	/* Set Feature */
+	fu_common_dump_raw (G_LOG_DOMAIN, "SetFeature", data, datasz);
+	if (!fu_synaptics_rmi_device_ioctl (self, HIDIOCSFEATURE(datasz), data, error)) {
+		g_prefix_error ("failed to SetFeature: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_rmi_device_get_feature (FuSynapticsRmiDevice *self,
+				     guint8 *data,
+				     guint datasz,
+				     GError **error)
+{
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	if (!fu_synaptics_rmi_device_ioctl (self, HIDIOCGFEATURE(datasz), data, error)) {
+		g_prefix_error ("failed to GetFeature: ");
+		return FALSE;
+	}
+	fu_common_dump_raw (G_LOG_DOMAIN, "GetFeature", data, datasz);
+	return TRUE;
+}
+#endif
+
+static gboolean
+fu_synaptics_rmi_device_setup (FuDevice *device, GError **error)
+{
+	FuDeviceClass *klass_device = FU_DEVICE_GET_CLASS (device);
+	FuSynapticsRmiDeviceClass *klass_rmi = FU_SYNAPTICS_RMI_DEVICE_GET_CLASS (device);
+	FuSynapticsRmiDevice *self = FU_SYNAPTICS_RMI_DEVICE (device);
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	guint16 addr;
+	guint16 prod_info_addr;
+	guint8 ds4_query_length = 0;
+	guint8 has_build_id_query = FALSE;
+	guint8 has_package_id_query = FALSE;
+	g_autofree gchar *bl_ver = NULL;
+	g_autofree gchar *board_id = NULL;
+	g_autofree gchar *fw_ver = NULL;
+	g_autofree gchar *instance_id = NULL;
+	g_autoptr(GByteArray) f01_basic = NULL;
+	g_autoptr(GByteArray) f01_db = NULL;
+	g_autoptr(GByteArray) f01_product_id = NULL;
+	g_autoptr(GByteArray) f34_ctrl = NULL;
+
+	/* read basic device information */
+	if (!fu_synaptics_rmi_device_set_rma_page (self, 0x00, error))
+		return FALSE;
+
+	/* read PDT */
+	if (!fu_synaptics_rmi_device_scan_pdt (self, error))
+		return FALSE;
+	priv->f01 = fu_synaptics_rmi_device_get_function (self, 0x01, error);
+	if (priv->f01 == NULL)
+		return FALSE;
+	addr = priv->f01->query_base;
+	f01_basic = fu_synaptics_rmi_device_read (self, addr, RMI_DEVICE_F01_BASIC_QUERY_LEN, error);
+	if (f01_basic == NULL) {
+		g_prefix_error (error, "failed to read the basic query: ");
+		return FALSE;
+	}
+	priv->manufacturer_id = f01_basic->data[0]; // FIXME: vendor_id?
+	priv->has_lts = f01_basic->data[1] & RMI_DEVICE_F01_QRY1_HAS_LTS;
+	priv->has_sensor_id = f01_basic->data[1] & RMI_DEVICE_F01_QRY1_HAS_SENSOR_ID;
+	priv->has_query42 = f01_basic->data[1] & RMI_DEVICE_F01_QRY1_HAS_PROPS_2;
+
+	/* use the product ID as an instance ID */
+	addr += 11;
+	f01_product_id = fu_synaptics_rmi_device_read (self, addr, RMI_PRODUCT_ID_LENGTH, error);
+	if (f01_product_id == NULL) {
+		g_prefix_error (error, "failed to read the product id: ");
+		return FALSE;
+	}
+	board_id = g_strndup ((const gchar *) f01_product_id->data, f01_product_id->len);
+	instance_id = g_strdup_printf ("SYNAPTICS_RMI\\%s", board_id);
+	fu_device_add_instance_id (device, instance_id);
```

Comment:
I think in a system that doesn't have a quirk, we can do better than what's showing for the name right now.  On the XPS 9380 I'm seeing:
`06CB:76AF` for `fu_device_get_name`.

We should probably have something in the udev plugin that does lookups for the common vendors and shows that when possible.  https://devicehunt.com/view/type/usb/vendor/06CB does come up as `Synaptics`.
The other option is to of course hardcode titles here unless overridden by quirk.

IE just use `Touchpad`

Vendor is showing up as "DELL08AF:00 (HIDRAW:0x06CB)".  I think that would be fine to just hardcode to "Synaptics" for stuff made by this plugin.

Reply:
I've hardcoded `Touchpad` now, thanks. We're using HID_NAME for the `name` -- is that wrong? Is there a better property or attribute we can read from the GUdevDevice?

Reply:
What about https://github.com/fwupd/fwupd/pull/1413/commits/f22bc0fde604493da3f814b3e26c51cb97cba4fa ?

Reply:
Yes that's better.